src/community/graticule/src/main/java/org/geotools/data/graticule/GraticuleDataStore.java

```diff
@@ -0,0 +1,388 @@
+package org.geotools.data.graticule;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.logging.Logger;
+import org.geotools.api.data.DataStore;
+import org.geotools.api.data.FeatureReader;
+import org.geotools.api.data.FeatureWriter;
+import org.geotools.api.data.LockingManager;
+import org.geotools.api.data.Query;
+import org.geotools.api.data.ServiceInfo;
+import org.geotools.api.data.SimpleFeatureSource;
+import org.geotools.api.data.Transaction;
+import org.geotools.api.feature.simple.SimpleFeature;
+import org.geotools.api.feature.simple.SimpleFeatureType;
+import org.geotools.api.feature.type.Name;
+import org.geotools.api.filter.Filter;
+import org.geotools.api.referencing.crs.CoordinateReferenceSystem;
+import org.geotools.data.CollectionFeatureReader;
+import org.geotools.data.DefaultServiceInfo;
+import org.geotools.data.graticule.gridsupport.LineFeatureBuilder;
+import org.geotools.feature.FeatureTypes;
+import org.geotools.feature.NameImpl;
+import org.geotools.feature.simple.SimpleFeatureTypeBuilder;
+import org.geotools.geometry.jts.ReferencedEnvelope;
+import org.geotools.grid.Lines;
+import org.geotools.grid.ortholine.LineOrientation;
+import org.geotools.grid.ortholine.OrthoLineDef;
+import org.locationtech.jts.geom.LineString;
+
+public class GraticuleDataStore implements DataStore {
+    static final Logger log = Logger.getLogger("GraticuleDataStore");
+    private final ReferencedEnvelope bounds;
+    private final List<Double> steps;
+    private final HashMap<Name, SimpleFeatureSource> sources = new HashMap<>();
+
+    private final ArrayList<Name> names = new ArrayList<>();
+
+    public GraticuleDataStore(ReferencedEnvelope env, List<Double> steps) {
+        this.steps = steps;
+        this.bounds = env;
+
+        int level = 0;
+        Collections.sort(steps);
+        for (double step : steps) {
+            Name name = new NameImpl(Double.toString(step).replace('.', '_'));
+            SimpleFeatureType schema = buildType(name, bounds.getCoordinateReferenceSystem());
+            log.fine("Creating graticule with name " + name);
+            names.add(name);
+
+            List<OrthoLineDef> lineDefs =
+                    Arrays.asList(
+                            // vertical (longitude) lines
+                            new OrthoLineDef(LineOrientation.VERTICAL, level, step),
+                            // horizontal (latitude) lines
+                            new OrthoLineDef(LineOrientation.HORIZONTAL, level, step));
+
+            // Specify vertex spacing to get "densified" polygons
+            double vertexSpacing = (bounds.getHeight() / 20); // should be dynamic
+            SimpleFeatureSource grid =
```

Comment:
We actually should not need to have pre-densified lines, the renderer has an on-the-fly densifier that takes into account the current resolution and  the amount of local deformation.
E.g., from the WMS admin panel:

![image](https://github.com/geoserver/geoserver/assets/325433/59beac10-3810-412d-8043-655f87b40c1c)

I don't remember if it's on by default, if not, we migth want to make it so on the developer branch.